### PR TITLE
v1_4_mBufferSetByteSlice

### PR DESCRIPTION
**Pull request by camilbancioiu**

### CODE REVIEW

File: arwen/elrondapi/manBufOps.go [hunk 1]
```diff
@@ -358,6 +364,42 @@ func v1_4_mBufferSetBytes(context unsafe.Pointer, mBufferHandle int32, dataOffse
...
+	if arwen.WithFault(err, context, runtime.ManagedBufferAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	if startingPosition < 0 || dataLength < 0 || int(startingPosition+dataLength) > len(bufferBytes) {
+		// does not fail execution if slice exceeds bounds
+		return 1
+	}
+
+	//TODO this loop should be faster than append() on slices, but needs benchmarking
```
Comment (sasurobert) [inline arwen/elrondapi/manBufOps.go:393]: Delete todo

File: arwen/elrondapi/manBufOps.go [hunk 2]
```diff
@@ -87,6 +88,11 @@ func ManagedBufferImports(imports *wasmer.Imports) (*wasmer.Imports, error) {
 		return nil, err
 	}
 
+	imports, err = imports.Append("mBufferSetByteSlice", v1_4_mBufferSetByteSlice, C.v1_4_mBufferSetByteSlice)
```
Comment (sasurobert) [inline arwen/elrondapi/manBufOps.go:91]: Need flag. Backward compatibility

Comment (sasurobert) [reply arwen/elrondapi/manBufOps.go:91]: Should be created on if flag is active. It is a little bit more complicated.

File: arwen/elrondapi/manBufOps.go (hunk 1 above)
Comment (camilbancioiu) [reply arwen/elrondapi/manBufOps.go:393]: Deleted.

File: arwen/elrondapi/elrondei.go [hunk 3]
```diff
@@ -914,6 +934,99 @@ func v1_4_getESDTTokenData(
...
+//export v1_4_validateTokenIdentifier
+func v1_4_validateTokenIdentifier(
+	context unsafe.Pointer,
+	tokenIdHandle int32,
+) int32 {
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.StorageLoad
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1018]: storageLoad ? that is a big cost. put something else here.

File: arwen/elrondapi/elrondei.go [hunk 4]
```diff
@@ -914,6 +934,99 @@ func v1_4_getESDTTokenData(
 	return int32(len(esdtData.Value.Bytes()))
 }
 
+//export v1_4_getESDTLocalRoles
+func v1_4_getESDTLocalRoles(context unsafe.Pointer, tokenIdHandle int32) int64 {
```
Comment (camilbancioiu) [inline arwen/elrondapi/elrondei.go:938]: This API method is being reworked due to a required fix.

File: arwen/elrondapi/elrondei.go (hunk 3 above)
Comment (sasurobert) [reply arwen/elrondapi/elrondei.go:1018]: Use getArgument cost.

File: arwen/elrondapi/elrondei.go [hunk 5]
```diff
@@ -914,6 +934,99 @@ func v1_4_getESDTTokenData(
...
+func validateToken(tokenID []byte) int32 {
+	tokenIDLen := len(tokenID)
+
+	if tokenIDLen < identifierMinLength || tokenIDLen > identifierMaxLength {
+		return 0
+	}
+
+	// ticker must be all uppercase alphanumeric
+	tickerLen := tokenIDLen - additionalRandomCharsLength
+
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:988]: delete empty line

File: arwen/elrondapi/elrondei.go (hunk 4 above)
Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go:938]: The code was expecting an incorrect type to be stored in the storage.

File: arwen/elrondapi/elrondei.go [hunk 6]
```diff
@@ -914,6 +934,99 @@ func v1_4_getESDTTokenData(
...
+	for i := 0; i < tickerLen-1; i++ {
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:989]: func isTickerValid(tickerName []byte) bool {
	if len(tickerName) < minLengthForTickerName || len(tickerName) > maxLengthForTickerName {
		return false
	}

	for _, ch := range tickerName {
		isBigCharacter := ch >= 'A' && ch <= 'Z'
		isNumber := ch >= '0' && ch <= '9'
		isReadable := isBigCharacter || isNumber
		if !isReadable {
			return false
		}
	}

	return true
}

File: arwen/elrondapi/elrondei.go [hunk 7]
```diff
@@ -914,6 +934,99 @@ func v1_4_getESDTTokenData(
...
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	storage := arwen.GetStorageContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	tokenID, err := managedType.GetBytes(tokenIdHandle)
+	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
+		return -1
+	}
+	key := []byte("ELRONDroleesdt" + string(tokenID))
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:948]: define the const for keyPrefix

File: arwen/elrondapi/elrondei.go [hunk 8]
```diff
@@ -914,6 +934,99 @@ func v1_4_getESDTTokenData(
...
+
+	data_buffer, usedCache := storage.GetStorage(key)
+	storage.UseGasForStorageLoad(
+		storageLoadName,
+		metering.GasSchedule().ElrondAPICost.StorageLoad,
+		len(data_buffer), usedCache)
+
+	result := int64(0)
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:956]: make a separate function out of lines 956-974 and you will be able to easily unit test

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: arwen/contexts/runtime.go [hunk 9]
```diff
@@ -533,11 +546,33 @@ func (context *runtimeContext) VerifyContractCode() error {
 		return err
 	}
 
+	if !context.flagUseDifferentGasCostForReadingCachedStorage.IsSet() {
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/runtime.go:549]: Done.

**Review by sasurobert — DISMISSED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by alyn509 — APPROVED** [review]

**Review by andrei-marinica — APPROVED** [review]

**Pull request merged**